evm/vm/base.py

```diff
@@ -252,15 +252,17 @@ def get_block_class(self):
     def get_block_by_header(self, block_header):
         return self.get_block_class().from_header(block_header, self.db)
 
-    def get_block_hash(self, block_number):
+    def get_ancestor_hash(self, block_number):
         """
-        For getting block hash for any block number in the the last 256 blocks.
+        Return the hash for the ancestor with the given number
         """
         ancestor_depth = self.block.number - block_number
-        if 1 <= ancestor_depth <= 256:
-            return lookup_block_hash(self.db, block_number)
-        else:
+        if ancestor_depth > 256 or ancestor_depth < 1:
             return b''
+        h = get_block_header_by_hash(self.db, self.block.header.parent_hash)
+        while h.block_number != block_number:
+            h = get_block_header_by_hash(self.db, h.parent_hash)
```

Comment:
Some functional cleverness that avoids this loop mutation, though questionable if it's a good idea.  Requires the [`toolz`](http://toolz.readthedocs.io/) library.

```python
from toolz.functoolz import pipe

header = pipe(
    self.block.header.parent_hash,
    *itertools.repeat(
        lambda h: get_block_header_by_hash(self.db, h.parent_hash), 
        ancestor_depth,
    ),
)
return header
```

Mostly just wanted to see what it looks like but lately I've been loving the `toolz` library.  Lots of really nice utilities in there that have helped with avoiding mutation in my code.